[RFC-0000] Introduce the Flux RFC template (#2085)
fluxcd/flux2: rfcs/README.md

# Flux RFCs

In many cases, new features and enhancements are proposed on [flux2/discussions](https://github.com/fluxcd/flux2/discussions).
A proposal is discussed in public by maintainers, contributors, users and other interested parties.
After some form of consensus is reached between participants, the proposed changes go through the
pull request process where the implementation details are reviewed, approved or rejected by maintainers.

Some proposals may be **substantial**, and for these we ask for a design process to be followed
so that all stakeholders can be confident about the direction Flux is evolving in.

> The "RFC" (request for comments) process is intended to provide a consistent and

Reviewer: Is there anything that describes the process?

Author: Added!

controlled path for substantial changes to enter Flux.

Examples of substantial changes:

- API additions (new kinds of resources, new relationships between existing APIs)
- API breaking changes (new required fields, field removals)
- Security related changes (Flux controllers permissions, tenant isolation and impersonation)
- Impactful UX changes (new required inputs to the bootstrap process)
- Drop capabilities (sunset an existing integration with an external service due to security concerns)

> ## RCF Template

Reviewer 2: ```suggestion
## RFC Template
```

```text
# RFC-NNNN Title

<!--
The title must be short and descriptive.
-->

## Summary

<!--
One paragraph explanation of the proposed feature or enhancement.
-->

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this RFC. Describe why the change is important and the benefits to users.
-->

### Goals

<!--

> List the specific goals of this RCF. What is it trying to achieve? How will we

Reviewer 2: ```suggestion
List the specific goals of this RFC. What is it trying to achieve? How will we
```

know that this has succeeded?
-->

### Non-Goals

<!--
What is out of scope for this RFC? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.
-->

### User Stories

<!--
Optional if existing discussions and/or issues are linked in the motivation section.
-->

### Alternatives

<!--

> List plausible alternatives to the proposal and
> motivate why the current proposal is superior.

Reviewer: ```suggestion
List plausible alternatives to the proposal and explain why the proposal is superior.

This is a good place to incorporate suggestions made during discussion of the RFC.
```

you're proposing, but should not include things like API designs or
implementation.
-->

### User Stories

<!--
Optional if existing discussions and/or issues are linked in the motivation section.
-->

### Alternatives

<!--
List plausible alternatives to the proposal and
motivate why the current proposal is superior.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs and code snippets.

The design details should answer the following questions:
- How can this feature be enabled / disabled?
- Does enabling the feature change any default behavior?
- Can the feature be disabled once it has been enabled?
- How can an operator determine if the feature is in use?
- Are there any drawbacks when enabling this feature?
-->

## Implementation History

<!--
Major milestones in the lifecycle of the RFC such as:
- The first Flux release where an initial version of the RFC was available.
- The version of Flux where the RFC graduated to general availability.
- The version of Flux where the RFC was retired or superseded.
-->
```

> The RFC template is heavily inspired by Rust RFCs and Kubernetes KEPs.

Reviewer 3: Is this of importance to the reader? (If done for being a good crediting fellow, I think putting it in the commit message would be sufficient).

After some form of consensus is reached between participants, the proposed changes go through the
pull request process where the implementation details are reviewed, approved or rejected by maintainers.

Some proposals may be **substantial**, and for these we ask for a design process to be followed
so that all stakeholders can be confident about the direction Flux is evolving in.

The "RFC" (request for comments) process is intended to provide a consistent and
controlled path for substantial changes to enter Flux.

Examples of substantial changes:

- API additions (new kinds of resources, new relationships between existing APIs)
- API breaking changes (new required fields, field removals)
- Security related changes (Flux controllers permissions, tenant isolation and impersonation)
- Impactful UX changes (new required inputs to the bootstrap process)
- Drop capabilities (sunset an existing integration with an external service due to security concerns)

## RFC Process

- Before submitting an RFC please discuss the proposal with the Flux community.
  Start a discussion on GitHub and ask for feedback at the weekly dev meeting.
  You must find a maintainer willing to sponsor the RFC.

> - Submit an RFC by opening a pull request using RFC-0000 as template.

Reviewer: Some RFCs might not follow the template exactly, e.g., if they document best practice rather than defining new behaviour. Perhaps we could allow for that here, or mention it in the template.

Author: I added a mention in the template. Thanks!

>   The sponsor will assign the PR to itself, will label the PR with `area/RFC` and

Reviewer: ```suggestion
  The sponsor will assign the PR to themselves, will label the PR with `area/RFC` and
```

  will request other maintainers to begin the review process.
- Integrate feedback by adding commits without overriding the history.
- At least two maintainers have to approve the proposal before it can be merged.

>   Approvers must be satisfied that an appropriate level of consensus has been reached.

Reviewer: :+1: We could say a bit more here to tie this to the governance advice for seeking consensus:

```suggestion
  Approvers must be satisfied that an [appropriate level of consensus](https://github.com/fluxcd/community/blob/main/GOVERNANCE.md#decision-guidelines) has been reached.
```

A proposal is discussed in public by maintainers, contributors, users and other interested parties.
After some form of consensus is reached between participants, the proposed changes go through the
pull request process where the implementation details are reviewed, approved or rejected by maintainers.

Some proposals may be **substantial**, and for these we ask for a design process to be followed
so that all stakeholders can be confident about the direction Flux is evolving in.

The "RFC" (request for comments) process is intended to provide a consistent and
controlled path for substantial changes to enter Flux.

Examples of substantial changes:

- API additions (new kinds of resources, new relationships between exiting APIs)
- API breaking changes (new required fields, field removals)
- Security related changes (Flux controllers permissions, tenant isolation and impersonation)
- Impactful UX changes (new required inputs to the bootstrap process)
- Drop capabilities (sunset an existing integration with an external service due to security concerns)

## RCF Template

```text

> # RCF-NNNN Title

Reviewer 3: How and when does the number get assigned?

Author: The number could be assigned at first review by a member of the oversight committee if the submitter gets it wrong.

-->
The title must be short and descriptive.

> <!--

Reviewer 4: The comment tags seem to be reversed here.
```md
<!--
The title must be short and descriptive.
--> 
```

## Summary

<!--
One paragraph explanation of the proposed feature or enhancement.
-->

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this RFC. Describe why the change is important and the benefits to users.
-->

### Goals

<!--
List the specific goals of this RCF. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

<!--
What is out of scope for this RFC? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.
-->

> 

Reviewer: I would like to see "Alternatives considered", where plausible alternatives to the proposal are described (at least briefly) and it is explained why the proposal is superior.

It's very useful to have this section, for a few reasons -- at least:

 - it encourages the author to think about alternatives, and why the proposal deserves to be proposed
 - it helps avoid [Chesterton's Fence](https://en.wikipedia.org/wiki/G._K._Chesterton#Chesterton's_fence) dilemmas, since the rationale for "this way and not another" is recorded
 - it can record the outcome of discussions about the RFC

Author: Added! Thanks

### User Stories

<!--
Optional if exiting discussions and/or issues are linked in the motivation section.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs and code snippets.

> The design details should answer the following questions:

Reviewer: .. but not _just_ these questions, right?

Author: Right! These are here for the submitter to anticipate some questions that could arise during the review process.
